compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/IrElement.kt

```diff
@@ -20,22 +20,18 @@ import org.jetbrains.kotlin.ir.visitors.IrElementTransformer
 import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
 import org.jetbrains.kotlin.ir.visitors.IrElementVisitor
 
-interface IrElement {
-    val startOffset: Int
-    val endOffset: Int
-
+// A facade for IrElement which is autogenerated
+interface IrElementBase {
```

Comment:
Yeah, I'll do it. Right now I'm not entirely sure that this interface is needed at all, as we can just generate its contents into `IrElement`.

Reply:
Yeah, these methods *could* be generated intristictly in IrTree.kt. However I'd like to avoid that in order to keep the tree generator clean. This way is should also be easier to later work on.
I added ability to write custom code to the generator to support cases where I did not see feasible way to work around it, but ideally I see it without them. Maybe the facade patter n should actually be carried out to them too.

Reply:
I'm leaning towards generating these methods in `IrElement` for the time being. This is just such a central place of the infrastructure that I'm wary that extra indirection here might cost something. Specifically, after such change methods `acceptChildren/transform/transformChildren` in `IrElement$DefaultImpls` now have a **checkcast to interface** `IrElementBase` and call to `IrElementBase$DefaultImpls`.

I can't confirm this suspicion based on our internal performance tests, but I don't trust them fully to be able to capture such regression because of unclear measurement error. So I'd prefer to stay on the safe side for now and keep `IrElement` exactly as in master. We can always change it later, after we either convince ourselves that the change is harmless, or maybe after we transition to using `-Xjvm-default=all` in the compiler (it's kind of in progress already).

Reply:
Ah yes, I forgot $DefaultImpls is a thing here on current abi target.
So actually, now I see `IrAbstractElement` should not be auto-gened. Then, base impls for `acceptChildren/transform/transformChildren` could be in it, giving old (if not better? there is no $DefaultImpls anymore) performance.

Reply:
Yeah, but `IrElement` still has to declare `accept/acceptChildren/transform/transformChildren` at least as abstract because there are a lot of call sites of those methods on values of type `IrElement` across the project right now. So the generator still needs to have custom code to generate them.

Most of these call sites are from visitor/transformer implementations though, so changing e.g. `IrElementVisitor.visitElement` to take `IrAbstractElement` instead of `IrElement` will allow to remove these methods from `IrElement` (and hopefully positively affect performance, as you mentioned). But that means changing almost all visitor implementations in all backends, and I'd like to avoid it right now.

So as a first step, I'll restore the `accept/acceptChildren/transform/transformChildren` methods in `IrElement`. Also, I'll probably make `IrAbstractElement` non-autogenerated, as you suggest, just because I've already done it and it somewhat simplifies the generator :). Then in the future, after merging the PR, I'll look into changing `visitElement`'s parameter type to `IrAbstractElement` and moving accept/transform methods from `IrElement` to `IrAbstractElement`.

Reply:
Well, I meant to just leave abstract declarations of these methods in `IrElementBase` or `IrElement` and move default impl to `IrAbstractElement`. That way there's no need to change any declaration to accept `IrAbstractElement`, although doing that way has the potential of further boosting perf somewhat. Not sure if all of this matters much on modern JVMs though.

Reply:
Indeed, I'll move the implementation of `transform` to `IrAbstractElement`, thanks!